### PR TITLE
3716 raster folder

### PR DESCRIPTION
**Pull request by sblack-usu**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [x] Automated Testing

- [x] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
1. Place two different rasters into a composite resource and ensure both are aggregated successfully.

### CODE REVIEW

**Comment by sblack-usu** [comment]
retest this, please

**Comment by sblack-usu** [comment]
@pkdash - Could you take a look at this?  Somehow multiple raster aggregations in the same folder was overlooked in the Composite Resource folderless aggregation.

**Comment by pkdash** [comment]
@sblack-usu Will have a look at this either today or tomorrow.

File: hs_file_types/models/raster.py [hunk 1]
```diff
@@ -416,46 +415,46 @@ def raster_file_validation(raster_file, resource, raster_folder=None):
...
 
-        if vrt_files:
+        if vrt_file:
+            raster_resource_files.extend([vrt_file])
+            create_vrt = False
             temp_dir = os.path.dirname(raster_file)
-            temp_vrt_file = utils.get_file_from_irods(vrt_files[0], temp_dir)
+            temp_vrt_file = utils.get_file_from_irods(vrt_file, temp_dir)
+            listed_tif_files = list_tif_files(vrt_file)
+            tif_files = [f for f in res_files if f.file_name in listed_tif_files]
```
Comment (pkdash) [inline hs_file_types/models/raster.py]: We should check that ```len(listed_tif_files) == len(tif_files)``` otherwise ```return validation_results```

Comment (sblack-usu) [reply hs_file_types/models/raster.py]: done

File: hs_file_types/models/raster.py [hunk 2]
```diff
@@ -525,33 +524,68 @@ def raster_file_validation(raster_file, resource, raster_folder=None):
...
+    Searches for a vrt_file that lists the supplied rater_file
+    :param raster_file: The raster file to find the associated vrt_file of
+    :param res_files: list of ResourceFiles in the the folder of raster_file
+    :return: The ResourceFile of a vrt file which lists the raster_file, None if not found
+    """
+    vrt_files = [f for f in res_files if f.extension.lower() == ".vrt"]
+    if vrt_files:
+        for vrt_file in vrt_files:
+            file_names_in_vrt = list_tif_files(vrt_file)
+            for vrt_ref_raster_name in file_names_in_vrt:
```
[the remark this answers is not in the document]
Comment (sblack-usu) [reply hs_file_types/models/raster.py]: great idea with the empty list, done

File: hs_file_types/models/raster.py [hunk 3]
```diff
@@ -525,33 +524,68 @@ def raster_file_validation(raster_file, resource, raster_folder=None):
...
+                if raster_file.endswith(vrt_ref_raster_name):
+                    return vrt_file
```
[the remark this answers is not in the document]
Comment (sblack-usu) [reply hs_file_types/models/raster.py]: Is it really invalid if two vrt files reference the same file?  This feels like an uber rare edge case.  

I guess our question should be is, how does hydroshare behave if two vrt files reference the same tif file?  Does the tif file show up in both aggregations?  Does it break everything?  I'll run through some tests.

File: hs_file_types/models/raster.py [hunk 4]
```diff
@@ -525,33 +524,68 @@ def raster_file_validation(raster_file, resource, raster_folder=None):
...
+        vrt_string = opened_vrt_file.read()
+        root = ET.fromstring(vrt_string)
+        file_names_in_vrt = [file_name.text for file_name in root.iter('SourceFilename')]
+        return file_names_in_vrt
+    return None
+
+
+def get_vrt_file(raster_file, res_files):
+    """
+    Searches for a vrt_file that lists the supplied rater_file
```
[the remark this answers is not in the document]
Comment (sblack-usu) [reply hs_file_types/models/raster.py]: fixed

**Comment by sblack-usu** [comment]
@pkdash - sorry for the delay in following up on your code review.  I addressed the questions I was confident in.  The only issue that remains to be resolved is your question to what happens if we have a tif file referenced by two vrt files.  I'd like to do some tests around this to see how it currently behaves.

File: hs_file_types/models/raster.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sblack-usu) [reply hs_file_types/models/raster.py]: ok, so ran through some tests with this scenario.  All that happened was that the tif file ended up in the most recently created aggregation.  The original aggregation no longer has a file, but still exists.  This is a problem.  i'll put in a validation check for this scenario.

**Comment by pkdash** [comment]
@sblack-usu No problem. Let me know when you want me to look at this again.

**Comment by sblack-usu** [comment]
@pkdash - could you take another look at this?

**Comment by pkdash** [comment]
@sblack-usu Will try to look at this tomorrow night.

**Pull request merged**